<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Modstrings • Modstrings</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Modstrings">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Modstrings</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Modstrings.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ModDNAString-alphabet.html">ModDNAString alphabet</a>
    </li>
    <li>
      <a href="../articles/ModRNAString-alphabet.html">ModRNAString alphabet</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Modstrings</h1>
                        <h4 class="author">Felix G.M. Ernst &amp; Denis L.J. Lafontaine</h4>
            
            <h4 class="date">2020-03-02</h4>
      
      
      <div class="hidden name"><code>Modstrings.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      Classes for DNA and RNA sequences containing modified nucleotides
    </div>
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Most nucleic acids, regardless of their being DNA or RNA, contain modified nucleotides, which enhances the normal function of encoding genetic information. They have usually a regulatory function and/or modify folding behavior and molecular interactions.</p>
<p>RNA are nearly always post-transcriptionally modified. Most prominent examples are of course ribsomal RNA (rRNA) and transfer RNA (tRNA), but in recent years mRNA was also discovered to be post-transcriptionally modified. In addition, many small and long non-coding RNAs are also modified.</p>
<p>In many resources, like the tRNAdb <span class="citation">(Jühling et al. 2009)</span> or the modomics database <span class="citation">(Boccaletto et al. 2018)</span>, modified nucleotides are repertoried. However in the Bioconductor context these information were not accessible, since they rely extensively on special characters in the RNA modification alphabet.</p>
<p>Therefore, the <code>ModRNAString</code> class was implemented extending the <code>BString</code> class from the <code>Biostrings</code> <span class="citation">(H. Pagès, P. Aboyoun, R. Gentleman, and S. DebRoy 2017)</span> package. It can store RNA sequences containing special characters of the RNA modification alphabet and thus can store location and identity of modifications. Functions for conversion to a tabular format are implemented as well.</p>
<p>The implemented classes inherit most of the functions from the parental <code>BString</code> class and it derivatives, which allows them to behave like the normal <code>XString</code> classes within the Bioconductor context. Most of the functionality is directly inherited and derived from the <code>Biostrings</code> package.</p>
<p>Since a DNA modification alphabet also exists, a <code>ModDNAString</code> class was implemented as well. For details on the available letters have a look at the <a href="ModRNAString-alphabet.html">RNA modification</a> and [DNA modification](ModDNAString-alphabet.html alphabet vignettes.</p>
</div>
<div id="creating-a-modrnastring-object" class="section level1">
<h1 class="hasAnchor">
<a href="#creating-a-modrnastring-object" class="anchor"></a>Creating a ModRNAString object</h1>
<p>In principle <code>ModRNAString</code> and <code>ModDNAString</code> objects can be created as any other <code>XString</code> object. However encoding issue will most certainly come into play, depending on the modification, the operation system and probably the R version. This is not a problem of how the data is internally used, but how the letter is transfered from the console to R and back.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(Modstrings)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(GenomicRanges)</a></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># This works</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">mr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ModRNAString.html">ModRNAString</a></span>(<span class="st">"ACGU7"</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co"># This might work on Linux, but does not on Windows</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="kw"><a href="../reference/ModRNAString.html">ModRNAString</a></span>(<span class="st">"ACGU≈"</span>)</a></code></pre></div>
<pre><code>## 5-letter ModRNAString object
## seq: ACGU≈</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># This cause a misinterpretation on Windows. Omega gets added as O. </span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co"># This modifys the information from yW-72 (7-aminocarboxypropylwyosine) to </span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="co"># m1I (1-methylinosine)</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="kw"><a href="../reference/ModRNAString.html">ModRNAString</a></span>(<span class="st">"ACGUΩ"</span>)</a></code></pre></div>
<pre><code>## 5-letter ModRNAString object
## seq: ACGUΩ</code></pre>
<p>To eliminate this issue the function <code>modifyNucleotide()</code> is implemented, which can use short names or the nomenclature of a modification to add it at the desired position.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw"><a href="../reference/alphabet.html">shortName</a></span>(<span class="kw"><a href="../reference/ModRNAString.html">ModRNAString</a></span>()))</a></code></pre></div>
<pre><code>## [1] "m1Am"    "m1Gm"    "m1Im"    "m1acp3Y" "m1A"     "m1G"</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw"><a href="../reference/alphabet.html">nomenclature</a></span>(<span class="kw"><a href="../reference/ModRNAString.html">ModRNAString</a></span>()))</a></code></pre></div>
<pre><code>## [1] "01A"   "01G"   "019A"  "1309U" "1A"    "1G"</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">r &lt;-<span class="st"> </span><span class="kw">RNAString</span>(<span class="st">"ACGUG"</span>)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">mr2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/modifyNucleotides.html">modifyNucleotides</a></span>(r,5L,<span class="st">"m7G"</span>)</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">mr2</a></code></pre></div>
<pre><code>## 5-letter ModRNAString object
## seq: ACGU7</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">mr3 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/modifyNucleotides.html">modifyNucleotides</a></span>(r,5L,<span class="st">"7G"</span>,<span class="dt">nc.type =</span> <span class="st">"nc"</span>)</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">mr3</a></code></pre></div>
<pre><code>## 5-letter ModRNAString object
## seq: ACGU7</code></pre>
<p>In addition, one can also use the <code><a href="../reference/alphabet.html">alphabet()</a></code> function and subset to the desired modifications.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">mr4 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ModRNAString.html">ModRNAString</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"ACGU"</span>,<span class="kw"><a href="../reference/alphabet.html">alphabet</a></span>(<span class="kw"><a href="../reference/ModRNAString.html">ModRNAString</a></span>())[33L]))</a>
<a class="sourceLine" id="cb14-2" data-line-number="2">mr4</a></code></pre></div>
<pre><code>## 5-letter ModRNAString object
## seq: ACGUB</code></pre>
</div>
<div id="streamlining-object-creation-and-modification" class="section level1">
<h1 class="hasAnchor">
<a href="#streamlining-object-creation-and-modification" class="anchor"></a>Streamlining object creation and modification</h1>
<p>To offer a more streamlined functionality, which can take more information as input, the function <code><a href="../reference/separate.html">combineIntoModstrings()</a></code> is implemented. It takes a <code>XString</code> object and a <code>GRanges</code> object with a <code>mod</code> column and returns a <code>ModString</code> object. The information in the <code>mod</code> column must match the short name or nomenclature of the particular modification of interest as returned by the <code><a href="../reference/alphabet.html">shortName()</a></code> or <code><a href="../reference/alphabet.html">nomenclature()</a></code> functions as seen above.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">gr &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span>(<span class="st">"1:5"</span>, <span class="dt">mod =</span> <span class="st">"m7G"</span>)</a>
<a class="sourceLine" id="cb16-2" data-line-number="2">mr5 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/separate.html">combineIntoModstrings</a></span>(r, gr)</a>
<a class="sourceLine" id="cb16-3" data-line-number="3">mr5</a></code></pre></div>
<pre><code>## 5-letter ModRNAString object
## seq: ACGU7</code></pre>
<p><code><a href="../reference/separate.html">combineIntoModstrings()</a></code> is also implemented for <code>ModStringSet</code> objects.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">rs &lt;-<span class="st"> </span><span class="kw">RNAStringSet</span>(<span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(r,r,r,r,r))</a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(rs) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"Sequence"</span>, <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(rs))</a>
<a class="sourceLine" id="cb18-3" data-line-number="3">gr2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span>(<span class="dt">seqnames =</span> <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(rs)[<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(1L,1L,2L,3L,3L,4L,5L,5L)],</a>
<a class="sourceLine" id="cb18-4" data-line-number="4">               <span class="dt">ranges =</span> <span class="kw">IRanges</span>(<span class="dt">start =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(4L,5L,5L,4L,5L,5L,4L,5L),</a>
<a class="sourceLine" id="cb18-5" data-line-number="5">                                <span class="dt">width =</span> 1L),</a>
<a class="sourceLine" id="cb18-6" data-line-number="6">               <span class="dt">mod =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"D"</span>,<span class="st">"m7G"</span>,<span class="st">"m7G"</span>,<span class="st">"D"</span>,<span class="st">"m7G"</span>,<span class="st">"m7G"</span>,<span class="st">"D"</span>,<span class="st">"m7G"</span>))</a>
<a class="sourceLine" id="cb18-7" data-line-number="7">gr2</a></code></pre></div>
<pre><code>## GRanges object with 8 ranges and 1 metadata column:
##        seqnames    ranges strand |         mod
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   [1] Sequence1         4      * |           D
##   [2] Sequence1         5      * |         m7G
##   [3] Sequence2         5      * |         m7G
##   [4] Sequence3         4      * |           D
##   [5] Sequence3         5      * |         m7G
##   [6] Sequence4         5      * |         m7G
##   [7] Sequence5         4      * |           D
##   [8] Sequence5         5      * |         m7G
##   -------
##   seqinfo: 5 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">mrs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/separate.html">combineIntoModstrings</a></span>(rs, gr2)</a>
<a class="sourceLine" id="cb20-2" data-line-number="2">mrs</a></code></pre></div>
<pre><code>##   A ModRNAStringSet instance of length 5
##     width seq                                               names               
## [1]     5 ACGD7                                             Sequence1
## [2]     5 ACGU7                                             Sequence2
## [3]     5 ACGD7                                             Sequence3
## [4]     5 ACGU7                                             Sequence4
## [5]     5 ACGD7                                             Sequence5</code></pre>
<p>The reverse operation is also available via the function <code><a href="../reference/separate.html">separate()</a></code>, which allows the positions of modifications to be transfered into a tabular format.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">gr3 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/separate.html">separate</a></span>(mrs)</a>
<a class="sourceLine" id="cb22-2" data-line-number="2">rs2 &lt;-<span class="st"> </span><span class="kw">RNAStringSet</span>(mrs)</a>
<a class="sourceLine" id="cb22-3" data-line-number="3">gr3</a></code></pre></div>
<pre><code>## GRanges object with 8 ranges and 1 metadata column:
##        seqnames    ranges strand |         mod
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   [1] Sequence1         4      + |           D
##   [2] Sequence1         5      + |         m7G
##   [3] Sequence2         5      + |         m7G
##   [4] Sequence3         4      + |           D
##   [5] Sequence3         5      + |         m7G
##   [6] Sequence4         5      + |         m7G
##   [7] Sequence5         4      + |           D
##   [8] Sequence5         5      + |         m7G
##   -------
##   seqinfo: 5 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1">rs2</a></code></pre></div>
<pre><code>## RNAStringSet object of length 5:
##     width seq                                               names               
## [1]     5 ACGUG                                             Sequence1
## [2]     5 ACGUG                                             Sequence2
## [3]     5 ACGUG                                             Sequence3
## [4]     5 ACGUG                                             Sequence4
## [5]     5 ACGUG                                             Sequence5</code></pre>
<p><code><a href="../reference/modifyNucleotides.html">modifyNucleotides()</a></code> and therefore also <code><a href="../reference/separate.html">combineIntoModstrings()</a></code> requires, that the nucleotides to be modified match the originating base for the modification. The next chunk fails, since the originating base for m7G is of course G.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1"><span class="kw"><a href="../reference/modifyNucleotides.html">modifyNucleotides</a></span>(r,4L,<span class="st">"m7G"</span>)</a></code></pre></div>
<pre><code>## Error: Modification type does not match the originating base:
##  U != G  for  m7G</code></pre>
<p>Calls for both functions check the sanity for this operation, so that the next bit is always <code>TRUE</code>.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">r &lt;-<span class="st"> </span><span class="kw">RNAString</span>(<span class="st">"ACGUG"</span>)</a>
<a class="sourceLine" id="cb28-2" data-line-number="2">mr2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/modifyNucleotides.html">modifyNucleotides</a></span>(r,5L,<span class="st">"m7G"</span>)</a>
<a class="sourceLine" id="cb28-3" data-line-number="3">r <span class="op">==</span><span class="st"> </span><span class="kw">RNAString</span>(mr2)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="comparing-modstring-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#comparing-modstring-objects" class="anchor"></a>Comparing ModString objects</h1>
<p><code>ModString</code> objects can be directly compared to <code>RNAString</code> or <code>DNAString</code> objects depending on the type (<code>ModRNA</code> to <code>RNA</code> and <code>ModDNA</code> to <code>DNA</code>).</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1">r <span class="op">==</span><span class="st"> </span><span class="kw"><a href="../reference/ModRNAString.html">ModRNAString</a></span>(r)</a></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1">r <span class="op">==</span><span class="st"> </span>mr</a></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb34-1" data-line-number="1">rs <span class="op">==</span><span class="st"> </span><span class="kw"><a href="../reference/ModStringSet.html">ModRNAStringSet</a></span>(rs)</a></code></pre></div>
<pre><code>## [1] TRUE TRUE TRUE TRUE TRUE</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1">rs <span class="op">==</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(mrs[1L<span class="op">:</span>3L],rs[4L<span class="op">:</span>5L])</a></code></pre></div>
<pre><code>## [1] FALSE FALSE FALSE  TRUE  TRUE</code></pre>
</div>
<div id="conversion-of-modstring-objects" class="section level1">
<h1 class="hasAnchor">
<a href="#conversion-of-modstring-objects" class="anchor"></a>Conversion of ModString objects</h1>
<p><code>ModString</code> objects can be converted into each other. However any conversion will remove any information on modifications and revert each nucleotide back to its originating nucleotide.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="kw">RNAString</span>(mr)</a></code></pre></div>
<pre><code>## 5-letter RNAString object
## seq: ACGUG</code></pre>
</div>
<div id="quality-scaled-modstring" class="section level1">
<h1 class="hasAnchor">
<a href="#quality-scaled-modstring" class="anchor"></a>Quality scaled ModString</h1>
<p>Quality information can be encoded alongside <code>ModString</code> objects by combining it with a <code>XStringQuality</code> object inside a <code>QualityScaledModStringSet</code> object. Two class are implemented: <code>QualityScaledModRNAStringSet</code> and <code>QualityScaledModDNAStringSet</code>. They are usable as expected from a <code>QualityScaledXStringSet</code> object.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1">qmrs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/QualityScaledModStringSet.html">QualityScaledModRNAStringSet</a></span>(mrs,</a>
<a class="sourceLine" id="cb40-2" data-line-number="2">                                     <span class="kw">PhredQuality</span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"!!!!h"</span>,<span class="st">"!!!!h"</span>,<span class="st">"!!!!h"</span>,</a>
<a class="sourceLine" id="cb40-3" data-line-number="3">                                                    <span class="st">"!!!!h"</span>,<span class="st">"!!!!h"</span>)))</a>
<a class="sourceLine" id="cb40-4" data-line-number="4">qmrs</a></code></pre></div>
<pre><code>##   A QualityScaledModRNAStringSet instance containing:
## 
##   A ModRNAStringSet instance of length 5
##     width seq                                               names               
## [1]     5 ACGD7                                             Sequence1
## [2]     5 ACGU7                                             Sequence2
## [3]     5 ACGD7                                             Sequence3
## [4]     5 ACGU7                                             Sequence4
## [5]     5 ACGD7                                             Sequence5
## 
## PhredQuality object of length 5:
##     width seq
## [1]     5 !!!!h
## [2]     5 !!!!h
## [3]     5 !!!!h
## [4]     5 !!!!h
## [5]     5 !!!!h</code></pre>
<p>They can also be constructed/deconstructed using the functions <code><a href="../reference/separate.html">combineIntoModstrings()</a></code> and <code><a href="../reference/separate.html">separate()</a></code> and use an additional metadata column named <code>quality</code>. For quality information to persist during construction, set the argument <code>with.qualities = TRUE</code>. If a <code>QualityScaledModStringSet</code> is used as an input to separate, the quality information are returned in the <code>quality column</code>. We choose to avoid clashes with the <code>score</code> column and not to recycle it.</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1">qgr &lt;-<span class="st"> </span><span class="kw"><a href="../reference/separate.html">separate</a></span>(qmrs)</a>
<a class="sourceLine" id="cb42-2" data-line-number="2">qgr</a></code></pre></div>
<pre><code>## GRanges object with 8 ranges and 2 metadata columns:
##        seqnames    ranges strand |         mod   quality
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt;
##   [1] Sequence1         4      + |           D        71
##   [2] Sequence1         5      + |         m7G         0
##   [3] Sequence2         5      + |         m7G        71
##   [4] Sequence3         4      + |           D         0
##   [5] Sequence3         5      + |         m7G        71
##   [6] Sequence4         5      + |         m7G         0
##   [7] Sequence5         4      + |           D        71
##   [8] Sequence5         5      + |         m7G        71
##   -------
##   seqinfo: 5 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1"><span class="kw"><a href="../reference/separate.html">combineIntoModstrings</a></span>(mrs,qgr, <span class="dt">with.qualities =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>##   A QualityScaledModRNAStringSet instance containing:
## 
##   A ModRNAStringSet instance of length 5
##     width seq                                               names               
## [1]     5 ACGD7                                             Sequence1
## [2]     5 ACGU7                                             Sequence2
## [3]     5 ACGD7                                             Sequence3
## [4]     5 ACGU7                                             Sequence4
## [5]     5 ACGD7                                             Sequence5
## 
## PhredQuality object of length 5:
##     width seq
## [1]     5 !!!h!
## [2]     5 !!!!h
## [3]     5 !!!!h
## [4]     5 !!!!!
## [5]     5 !!!hh</code></pre>
</div>
<div id="saving-and-reading-modstring-objects-tofrom-file" class="section level1">
<h1 class="hasAnchor">
<a href="#saving-and-reading-modstring-objects-tofrom-file" class="anchor"></a>Saving and reading ModString objects to/from file</h1>
<p>The nucleotide sequences with modifications can be saved to a <code>fasta</code> or <code>fastq</code> file using the functions <code><a href="../reference/ModStringSet-io.html">writeModStringSet()</a></code>. Reading of these files is achieved using <code><a href="../reference/ModStringSet-io.html">readModRNAStringSet()</a></code> or <code><a href="../reference/ModStringSet-io.html">readModDNAStringSet()</a></code>. In case of <code>fastq</code> files, the sequences can be automatically read as a <code>QualityScaledModRNAStringSet</code> using <code><a href="../reference/QualityScaledModStringSet.html">readQualityScaledModRNAStringSet()</a></code> function.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1"><span class="kw"><a href="../reference/ModStringSet-io.html">writeModStringSet</a></span>(mrs, <span class="dt">file =</span> <span class="st">"test.fasta"</span>)</a>
<a class="sourceLine" id="cb46-2" data-line-number="2"><span class="co"># note the different function name. Otherwise empty qualities will be written</span></a>
<a class="sourceLine" id="cb46-3" data-line-number="3"><span class="kw"><a href="../reference/QualityScaledModStringSet.html">writeQualityScaledModStringSet</a></span>(qmrs, <span class="dt">file =</span> <span class="st">"test.fastq"</span>)</a>
<a class="sourceLine" id="cb46-4" data-line-number="4">mrs2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ModStringSet-io.html">readModRNAStringSet</a></span>(<span class="st">"test.fasta"</span>, <span class="dt">format =</span> <span class="st">"fasta"</span>)</a>
<a class="sourceLine" id="cb46-5" data-line-number="5">mrs2</a></code></pre></div>
<pre><code>##   A ModRNAStringSet instance of length 5
##     width seq                                               names               
## [1]     5 ACGD7                                             Sequence1
## [2]     5 ACGU7                                             Sequence2
## [3]     5 ACGD7                                             Sequence3
## [4]     5 ACGU7                                             Sequence4
## [5]     5 ACGD7                                             Sequence5</code></pre>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" data-line-number="1">qmrs2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/QualityScaledModStringSet.html">readQualityScaledModRNAStringSet</a></span>(<span class="st">"test.fastq"</span>)</a>
<a class="sourceLine" id="cb48-2" data-line-number="2">qmrs2</a></code></pre></div>
<pre><code>##   A QualityScaledModRNAStringSet instance containing:
## 
##   A ModRNAStringSet instance of length 5
##     width seq                                               names               
## [1]     5 ACGD7                                             Sequence1
## [2]     5 ACGU7                                             Sequence2
## [3]     5 ACGD7                                             Sequence3
## [4]     5 ACGU7                                             Sequence4
## [5]     5 ACGD7                                             Sequence5
## 
## PhredQuality object of length 5:
##     width seq
## [1]     5 !!!!h
## [2]     5 !!!!h
## [3]     5 !!!!h
## [4]     5 !!!!h
## [5]     5 !!!!h</code></pre>
<p>Since these functions are specifically designed to work with the modified nucleotides within the sequence, they are slower than the analogous functions from the <code>Biostrings</code> package. This is the result of a purely R based implementation, whereas <code>Biostrings</code> functions are spead up through a C backend. This is a potential improvement for future developments, but currently special sequence files are limited, so it is not a priority.</p>
</div>
<div id="pattern-matching" class="section level1">
<h1 class="hasAnchor">
<a href="#pattern-matching" class="anchor"></a>Pattern matching</h1>
<p>Pattern matching is implemented as well as expected for <code>XString</code> objects.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb50-1" data-line-number="1"><span class="kw">matchPattern</span>(<span class="st">"U7"</span>,mr)</a></code></pre></div>
<pre><code>## Views on a 5-letter ModRNAString subject
## subject: ACGU7
## views:
##       start end width
##   [1]     4   5     2 [U7]</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" data-line-number="1"><span class="kw">vmatchPattern</span>(<span class="st">"D7"</span>,mrs)</a></code></pre></div>
<pre><code>## MIndex object of length 5
## $Sequence1
## IRanges object with 1 range and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         4         5         2
## 
## $Sequence2
## IRanges object with 0 ranges and 0 metadata columns:
##        start       end     width
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## 
## $Sequence3
## IRanges object with 1 range and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         4         5         2
## 
## $Sequence4
## IRanges object with 0 ranges and 0 metadata columns:
##        start       end     width
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
## 
## $Sequence5
## IRanges object with 1 range and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         4         5         2</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" data-line-number="1">mrl &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(mrs)</a>
<a class="sourceLine" id="cb54-2" data-line-number="2"><span class="kw">matchLRPatterns</span>(<span class="st">"7ACGU"</span>,<span class="st">"U7ACG"</span>,100L,mrl)</a></code></pre></div>
<pre><code>## Views on a 25-letter ModRNAString subject
## subject: ACGD7ACGU7ACGD7ACGU7ACGD7
## views:
##       start end width
##   [1]     5  23    19 [7ACGU7ACGD7ACGU7ACG]</code></pre>
</div>
<div id="future-development" class="section level1">
<h1 class="hasAnchor">
<a href="#future-development" class="anchor"></a>Future development</h1>
<p>In principle post-translational modifications of proteins could also be implemented. However, a one letter alphabet of post-translational modifications must be developed first. If you are already aware of such an alphabet and want to use it in a Bioconductor context, let us know.</p>
</div>
<div id="import-example" class="section level1">
<h1 class="hasAnchor">
<a href="#import-example" class="anchor"></a>Import example</h1>
<p>This is a quick example showing how sequence information containing modified nucleotides can be imported into an R session using the <code>Modstrings</code> package. The file needs to be UTF-8 encoded.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" data-line-number="1"><span class="co"># read the lines</span></a>
<a class="sourceLine" id="cb56-2" data-line-number="2">test &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/readLines.html">readLines</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>,<span class="st">"test.fasta"</span>,<span class="dt">package =</span> <span class="st">"Modstrings"</span>),</a>
<a class="sourceLine" id="cb56-3" data-line-number="3">                      <span class="dt">encoding =</span> <span class="st">"UTF-8"</span>)</a>
<a class="sourceLine" id="cb56-4" data-line-number="4"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(test,2L)</a></code></pre></div>
<pre><code>## [1] "&gt; tRNA | Ala | AGC | Saccharomyces cerevisiae | cytosolic"                                          
## [2] "-GGGCGUGUKGCGUAGDC-GGD--AGCGCRCUCCCUUIGCOPGGGAGAG-------------------GDCUCCGGTPCGAUUCCGGACUCGUCCACCA"</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" data-line-number="1"><span class="co"># keep every second line as sequence, the other one as name</span></a>
<a class="sourceLine" id="cb58-2" data-line-number="2">names &lt;-<span class="st"> </span>test[<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq.int</a></span>(<span class="dt">from =</span> 1L, <span class="dt">to =</span> 104L, <span class="dt">by =</span> 2L)]</a>
<a class="sourceLine" id="cb58-3" data-line-number="3">seq &lt;-<span class="st"> </span>test[<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq.int</a></span>(<span class="dt">from =</span> 2L, <span class="dt">to =</span> 104L, <span class="dt">by =</span> 2L)]</a>
<a class="sourceLine" id="cb58-4" data-line-number="4"><span class="co"># sanitize input. This needs to be adapt to the individual case</span></a>
<a class="sourceLine" id="cb58-5" data-line-number="5">names &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">" "</span>,<span class="st">"_"</span>,</a>
<a class="sourceLine" id="cb58-6" data-line-number="6">              <span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"&gt; "</span>,<span class="st">""</span>,</a>
<a class="sourceLine" id="cb58-7" data-line-number="7">                   <span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">" </span><span class="ch">\\</span><span class="st">| "</span>,<span class="st">"-"</span>,</a>
<a class="sourceLine" id="cb58-8" data-line-number="8">                        names)))</a>
<a class="sourceLine" id="cb58-9" data-line-number="9">seq &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"-"</span>,<span class="st">""</span>,<span class="kw"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span>(<span class="st">"_"</span>,<span class="st">""</span>,seq))</a>
<a class="sourceLine" id="cb58-10" data-line-number="10"><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(seq) &lt;-<span class="st"> </span>names</a></code></pre></div>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" data-line-number="1"><span class="co"># sanitize special characters to Modstrings equivalent</span></a>
<a class="sourceLine" id="cb59-2" data-line-number="2">seq &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sanitizeInput.html">sanitizeFromModomics</a></span>(seq)</a>
<a class="sourceLine" id="cb59-3" data-line-number="3">seq &lt;-<span class="st"> </span><span class="kw"><a href="../reference/ModStringSet.html">ModRNAStringSet</a></span>(seq)</a>
<a class="sourceLine" id="cb59-4" data-line-number="4">seq</a></code></pre></div>
<pre><code>##   A ModRNAStringSet instance of length 52
##      width seq                                              names               
##  [1]    76 GGGCGUGUKGCGUAGDCGGDAGC…TPCGAUUCCGGACUCGUCCACCA  tRNA-Ala-AGC-Saccha…
##  [2]    75 GCUCGCGUKLCGUAADGGCAACG…TPCG"CCCCCAUCGUGAGUGCCA  tRNA-Arg-UCU-Saccha…
##  [3]    76 PUCCUCGUKLCCCAADGGDCACG…TPCA"GUCCUGGCGGGGAAGCCA  tRNA-Arg-ICG-Saccha…
##  [4]    77 GACUCCAUGLCCAAGDDGGDDAA…TPCA"CCCUCACUGGGGUCGCCA  tRNA-Asn-GUU-Saccha…
##  [5]    75 UCCGUGAUAGUUPAADGGDCAGA…TPCAAUUCCCCGUCGCGGAGCCA  tRNA-Asp-GUC-Saccha…
##  ...   ... ...
## [48]    76 GCUCUCUUAGCUUAADGGDUAAA…TPCAAAUCAUGGAGAGAGUACCA  tRNA-Arg-NCU-Saccha…
## [49]    90 GGAUGGUUGACUGAGDGGDUUAA…TPCAAAUCCUACAUCAUCCGCCA  tRNA-Ser-UGA-Saccha…
## [50]    90 GGAUGGUUGACUGAGDGGDUUAA…TPCAAAUCCUACAUCAUCCGCCA  tRNA-Ser-UGA-Saccha…
## [51]    73 GUAAAUAUAAUUUAADGGDAAAA…TPCAAAUCUUAGUAUUUACACCA  tRNA-Thr-UAG-Saccha…
## [52]    74 AAGGAUAUAGUUUAADGGDAAAA…TPCGAAUCUCUUUAUCCUUGCCA  tRNA-Trp-!CA-Saccha…</code></pre>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" data-line-number="1"><span class="co"># convert the contained modifications into a tabular format</span></a>
<a class="sourceLine" id="cb61-2" data-line-number="2"><span class="kw"><a href="../reference/separate.html">separate</a></span>(seq)</a></code></pre></div>
<pre><code>## GRanges object with 567 ranges and 1 metadata column:
##                                                    seqnames    ranges strand |
##                                                       &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; |
##     [1]     tRNA-Ala-AGC-Saccharomyces_cerevisiae-cytosolic         9      + |
##     [2]     tRNA-Ala-AGC-Saccharomyces_cerevisiae-cytosolic        16      + |
##     [3]     tRNA-Ala-AGC-Saccharomyces_cerevisiae-cytosolic        20      + |
##     [4]     tRNA-Ala-AGC-Saccharomyces_cerevisiae-cytosolic        26      + |
##     [5]     tRNA-Ala-AGC-Saccharomyces_cerevisiae-cytosolic        34      + |
##     ...                                                 ...       ...    ... .
##   [563] tRNA-Trp-!CA-Saccharomyces_cerevisiae-mitochondrial        33      + |
##   [564] tRNA-Trp-!CA-Saccharomyces_cerevisiae-mitochondrial        36      + |
##   [565] tRNA-Trp-!CA-Saccharomyces_cerevisiae-mitochondrial        38      + |
##   [566] tRNA-Trp-!CA-Saccharomyces_cerevisiae-mitochondrial        52      + |
##   [567] tRNA-Trp-!CA-Saccharomyces_cerevisiae-mitochondrial        53      + |
##                 mod
##         &lt;character&gt;
##     [1]         m1G
##     [2]           D
##     [3]           D
##     [4]       m2,2G
##     [5]           I
##     ...         ...
##   [563]      cmnm5U
##   [564]          xA
##   [565]           Y
##   [566]         m5U
##   [567]           Y
##   -------
##   seqinfo: 47 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="sessioninfo" class="section level1">
<h1 class="hasAnchor">
<a href="#sessioninfo" class="anchor"></a>Sessioninfo</h1>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</a></code></pre></div>
<pre><code>## R Under development (unstable) (2020-03-02 r77886)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 16.04.6 LTS
## 
## Matrix products: default
## BLAS:   /home/travis/R-bin/lib/R/lib/libRblas.so
## LAPACK: /home/travis/R-bin/lib/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
## [1] GenomicRanges_1.39.2 GenomeInfoDb_1.23.13 Modstrings_1.3.9    
## [4] Biostrings_2.55.4    XVector_0.27.0       IRanges_2.21.3      
## [7] S4Vectors_0.25.12    BiocGenerics_0.33.0  BiocStyle_2.15.6    
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.3                 assertive.data_0.0-3      
##  [3] compiler_4.0.0             BiocManager_1.30.10       
##  [5] assertive.files_0.0-2      assertive.data.us_0.0-2   
##  [7] assertive.properties_0.0-4 bitops_1.0-6              
##  [9] tools_4.0.0                zlibbioc_1.33.1           
## [11] digest_0.6.25              assertive_0.3-5           
## [13] evaluate_0.14              memoise_1.1.0             
## [15] assertive.base_0.0-7       rlang_0.4.5               
## [17] yaml_2.2.1                 pkgdown_1.4.1             
## [19] xfun_0.12                  GenomeInfoDbData_1.2.2    
## [21] stringr_1.4.0              knitr_1.28                
## [23] assertive.sets_0.0-3       assertive.matrices_0.0-2  
## [25] assertive.datetimes_0.0-2  assertive.strings_0.0-3   
## [27] assertive.types_0.0-3      desc_1.2.0                
## [29] fs_1.3.1                   rprojroot_1.3-2           
## [31] assertive.code_0.0-3       R6_2.4.1                  
## [33] rmarkdown_2.1              bookdown_0.17             
## [35] magrittr_1.5               assertive.models_0.0-2    
## [37] assertive.numbers_0.0-2    codetools_0.2-16          
## [39] backports_1.1.5            htmltools_0.4.0           
## [41] MASS_7.3-51.5              assertthat_0.2.1          
## [43] stringi_1.4.6              assertive.data.uk_0.0-2   
## [45] RCurl_1.98-1.1             assertive.reflection_0.0-4
## [47] crayon_1.3.4</code></pre>
<p><a name="References"></a></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Boccaletto-2018">
<p>Boccaletto, Pietro, Magdalena A. Machnicka, Elzbieta Purta, Pawel Piatkowski, Blazej Baginski, Tomasz K. Wirecki, Valérie de Crécy-Lagard, et al. 2018. “MODOMICS: A Database of Rna Modification Pathways. 2017 Update.” <em>Nucleic Acids Research</em> 46 (D1): D303–D307. <a href="https://doi.org/10.1093/nar/gkx1030" class="uri">https://doi.org/10.1093/nar/gkx1030</a>.</p>
</div>
<div id="ref-Pages-2017">
<p>H. Pagès, P. Aboyoun, R. Gentleman, and S. DebRoy. 2017. “Biostrings.” Bioconductor. <a href="https://doi.org/10.18129/B9.bioc.Biostrings" class="uri">https://doi.org/10.18129/B9.bioc.Biostrings</a>.</p>
</div>
<div id="ref-Juehling-2009">
<p>Jühling, Frank, Mario Mörl, Roland K. Hartmann, Mathias Sprinzl, Peter F. Stadler, and Joern Pütz. 2009. “TRNAdb 2009: Compilation of tRNA Sequences and tRNA Genes.” <em>Nucleic Acids Research</em> 37: D159–D162. <a href="https://doi.org/10.1093/nar/gkn772" class="uri">https://doi.org/10.1093/nar/gkn772</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#creating-a-modrnastring-object">Creating a ModRNAString object</a></li>
      <li><a href="#streamlining-object-creation-and-modification">Streamlining object creation and modification</a></li>
      <li><a href="#comparing-modstring-objects">Comparing ModString objects</a></li>
      <li><a href="#conversion-of-modstring-objects">Conversion of ModString objects</a></li>
      <li><a href="#quality-scaled-modstring">Quality scaled ModString</a></li>
      <li><a href="#saving-and-reading-modstring-objects-tofrom-file">Saving and reading ModString objects to/from file</a></li>
      <li><a href="#pattern-matching">Pattern matching</a></li>
      <li><a href="#future-development">Future development</a></li>
      <li><a href="#import-example">Import example</a></li>
      <li><a href="#sessioninfo">Sessioninfo</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Felix G.M. Ernst, Denis L.J. Lafontaine.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
